### PR TITLE
Bug 1831853 - Migrate Open in app feature code from Java to Kotlin

### PR DESCRIPTION
**Pull request by mcarare**
### Pull Request checklist

- [x] **Quality**: This PR builds and passes detekt/ktlint checks (A pre-push hook is recommended)
- [x] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [x] **Changelog**: This PR includes [a changelog entry](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/changelog.md) or does not need one
- [x] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/firefox-android/blob/main/docs/shared/android/accessibility_guide.md) or does not include any user facing features

### After merge
- **Breaking Changes**: If this is a breaking Android Components change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

### To download an APK when reviewing a PR (after all CI tasks finished running):
1. Click on `Checks` at the top of the PR page.
2. Click on the `firefoxci-taskcluster` group on the left to expand all tasks.
3. Click on the `build-apk-{fenix,focus,klar}-debug` task you're interested in.
4. Click on `View task in Taskcluster` in the new `DETAILS` section.
5. The APK links should be on the right side of the screen, named for each CPU architecture.

### GitHub Automation
https://bugzilla.mozilla.org/show_bug.cgi?id=1831853

### CODE REVIEW

File: focus-android/app/src/main/java/org/mozilla/focus/open/AppAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+     * Interface for setting a listener for an [App] item.
+     */
+    interface OnAppSelectedListener {
+        /**
+         * Action to be performed on an item from [AppAdapter] list being clicked.
+         */
+        fun onAppSelected(app: App)
+    }
+
+    private val apps: List<App>
```
Comment (t-p-white) [inline focus-android/app/src/main/java/org/mozilla/focus/open/AppAdapter.kt:52]:
```suggestion
    private val apps: List<App> = infoArray.map { App(context, it) }
        .sortedWith { app1, app2 -> app1.label.compareTo(app2.label) }
```

File: focus-android/app/src/main/java/org/mozilla/focus/open/AppAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+    private val store: App?
```
Comment (t-p-white) [inline focus-android/app/src/main/java/org/mozilla/focus/open/AppAdapter.kt:53]:
```suggestion
    private val store: App?  = if (store != null) App(context, store) else null
```

File: focus-android/app/src/main/java/org/mozilla/focus/open/AppViewHolder.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+    private fun createListenerWrapper(
+        app: AppAdapter.App,
+        listener: OnAppSelectedListener?,
+    ): View.OnClickListener {
+        return View.OnClickListener { listener?.onAppSelected(app) }
+    }
+
+    companion object {
+        const val LAYOUT_ID = R.layout.item_app
```
Comment (t-p-white) [inline focus-android/app/src/main/java/org/mozilla/focus/open/AppViewHolder.kt:43]: I see a `NonConstantResourceId` lint warning here?

File: focus-android/app/src/main/java/org/mozilla/focus/open/OpenWithFragment.kt [hunk 4]
```diff
@@ -0,0 +1,147 @@
+/* -*- Mode: Java; c-basic-offset: 4; tab-width: 4; indent-tabs-mode: nil; -*-
```
Comment (t-p-white) [inline focus-android/app/src/main/java/org/mozilla/focus/open/OpenWithFragment.kt:1]:
```suggestion
/* 
```

File: focus-android/app/src/main/java/org/mozilla/focus/open/OpenWithFragment.kt [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+                val width =
+                    context.resources.getDimensionPixelSize(R.dimen.tablet_bottom_sheet_width)
+                val window = window
+                window?.setLayout(width, ViewGroup.LayoutParams.MATCH_PARENT)
+            }
+        }
+
+        override fun setContentView(view: View) {
+            super.setContentView(view)
+            this.contentView = view
```
Comment (t-p-white) [inline focus-android/app/src/main/java/org/mozilla/focus/open/OpenWithFragment.kt:89]:
```suggestion
           contentView = view
```

**Comment by t-p-white** [comment]
N.B The conversion has created some large blocks of code that it may be useful to manually break up, for example OpenWithFragment onAppSelected is a single block of code

File: focus-android/app/src/main/java/org/mozilla/focus/open/AppViewHolder.kt (hunk 3 above)
Comment (mcarare) [reply focus-android/app/src/main/java/org/mozilla/focus/open/AppViewHolder.kt:43]: Yeah, that is something we will need to change when upgrading to Gradle 8, and not just here.
For now, it's just a heads-up.

File: focus-android/app/src/main/java/org/mozilla/focus/open/AppAdapter.kt (hunk 1 above)
Comment (mcarare) [reply focus-android/app/src/main/java/org/mozilla/focus/open/AppAdapter.kt:52]: Thanks for the suggestion. I will keep initialisation in the init block, for visibility, but I will use the mapping suggestion, it works better! TY!

**Comment by nonremittal** [comment]
Does this mean that this feature is returning?

61745881de5f952d5fe623b667b34ab1b510d5cc_2_230x500

This was incredibly useful. If I used an external app for a particular site, I could find this in the menu (not a pop-up to install the _Official App For a Better Experience!_). Say for example you have a video player for specific URLs.

It was unobtrusive and meant I could navigate to a page, then use a program/app of my choice from the main menu. There was even an 'Install' option (guessing _for_ the official app).

**Pull request merged**